File: src/main/java/org/vors/pairbot/service/KeyboardService.java
package org.vors.pairbot.service;

import com.google.common.base.Joiner;
import org.apache.commons.lang3.BooleanUtils;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageReplyMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton;
import org.vors.pairbot.constant.Callback;
import org.vors.pairbot.model.Event;
import org.vors.pairbot.model.Participant;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.vors.pairbot.constant.BotConstants.CALLBACK_DATA_SEPARATOR;

@Component
public class KeyboardService {
    public InlineKeyboardMarkup getStartKeyboard() {
        return getOneRowKeyboard(button("New team", Callback.NEW_TEAM.toString()));
    }

    public InlineKeyboardMarkup getInviteKeyboard(Event event) {
        return getOneRowKeyboard(
                button(
                        "Decline",
                        Joiner.on(CALLBACK_DATA_SEPARATOR).join(Callback.ACCEPT_DECLINE.toString(), event.getPk(), Boolean.FALSE)
                ),
                button(
                        "Accept",
                        Joiner.on(CALLBACK_DATA_SEPARATOR).join(Callback.ACCEPT_DECLINE.toString(), event.getPk(), Boolean.TRUE)
                ));
    }

    public InlineKeyboardMarkup acceptedInviteKeyboard(Participant participant, Event event) {
        if (participant.isAccepted() != null || BooleanUtils.isFalse(event.getAccepted())) {
            return getRemoveKeyboardMarkup();
        } else {
            return getInviteKeyboard(event);
        }
    }

    public EditMessageReplyMarkup getRemoveKeyboard(Long chatId, Integer messageId) {
        return new EditMessageReplyMarkup()
                .setChatId(chatId)
                .setMessageId(messageId)
                .setReplyMarkup(getRemoveKeyboardMarkup());
    }

    public InlineKeyboardMarkup getRemoveKeyboardMarkup() {
        return getOneRowKeyboard();
    }

    private InlineKeyboardButton button(String label, String callbackData) {
        return callbackButton(label, callbackData);
    }

    private InlineKeyboardButton callbackButton(String text, String callbackData) {
        return new InlineKeyboardButton()
                .setText(text)
                .setCallbackData(callbackData);
    }

    private InlineKeyboardButton createLinkButton(String text, String url) {
        return new InlineKeyboardButton()
                .setText(text)
                .setUrl(url);
    }

    private InlineKeyboardMarkup getOneRowKeyboard(InlineKeyboardButton... buttons) {
        return getMultiRowKeyboard(new ArrayList<>(Collections.singletonList(row(buttons))));
    }

    private InlineKeyboardMarkup getMultiRowKeyboard(List<List<InlineKeyboardButton>> rows) {
        return new InlineKeyboardMarkup().setKeyboard(rows);
    }

    private List<InlineKeyboardButton> row(InlineKeyboardButton... buttons) {
        return new ArrayList<>(Arrays.asList(buttons));
    }

}

File: src/main/java/org/vors/pairbot/service/KeyboardService.kt
package org.vors.pairbot.service

import com.google.common.base.Joiner
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageReplyMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.ReplyKeyboard
import org.telegram.telegrambots.meta.api.objects.replykeyboard.ReplyKeyboardRemove
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton
import org.vors.pairbot.constant.Callback
import org.vors.pairbot.model.Event
import org.vors.pairbot.model.Participant

import java.util.ArrayList
import java.util.Arrays

import org.vors.pairbot.constant.BotConstants.CALLBACK_DATA_SEPARATOR
import org.vors.pairbot.model.ParticipantStatus.DECLINED
import org.vors.pairbot.model.ParticipantStatus.NO_RESPONSE

@Component
class KeyboardService {
    val startKeyboard: InlineKeyboardMarkup
        get() = getOneRowKeyboard(button("New team", Callback.NEW_TEAM.toString()))

    val removeKeyboardMarkup: InlineKeyboardMarkup
        get() = getOneRowKeyboard()

    fun getInviteKeyboard(event: Event): InlineKeyboardMarkup {
        return getOneRowKeyboard(
                button(
                        "Decline",
                        Joiner.on(CALLBACK_DATA_SEPARATOR).join(Callback.ACCEPT_DECLINE.toString(), event.pk, java.lang.Boolean.FALSE)
                ),
                button(
                        "Accept",
                        Joiner.on(CALLBACK_DATA_SEPARATOR).join(Callback.ACCEPT_DECLINE.toString(), event.pk, java.lang.Boolean.TRUE)
                ))
    }

    fun acceptedInviteKeyboard(participant: Participant, event: Event): InlineKeyboardMarkup {
        return if (participant.status != NO_RESPONSE || cancelled(event)) {
            removeKeyboardMarkup
        } else {
            getInviteKeyboard(event)
        }
    }

    private fun cancelled(event: Event): Boolean {
        return event.participants.any { it.status == DECLINED }
    }

    fun getRemoveKeyboard(chatId: String?, messageId: Int?): EditMessageReplyMarkup {
        return EditMessageReplyMarkup.builder()
                .chatId(chatId!!)
                .messageId(messageId)
                .replyMarkup(removeKeyboardMarkup)
                .build()
    }

    private fun button(label: String, callbackData: String): InlineKeyboardButton {
        return callbackButton(label, callbackData)
    }

    private fun callbackButton(text: String, callbackData: String): InlineKeyboardButton {
        return InlineKeyboardButton.builder()
                .text(text)
                .callbackData(callbackData)
                .build()
    }

    private fun createLinkButton(text: String, url: String): InlineKeyboardButton {
        return InlineKeyboardButton.builder()
                .text(text)
                .url(url)
                .build()
    }

    private fun getOneRowKeyboard(vararg buttons: InlineKeyboardButton): InlineKeyboardMarkup {
        return getMultiRowKeyboard(ArrayList(listOf(row(*buttons))))
    }

    private fun getMultiRowKeyboard(rows: List<List<InlineKeyboardButton>>): InlineKeyboardMarkup {
        return InlineKeyboardMarkup.builder()
                .keyboard(rows)
                .build()
    }

    private fun row(vararg buttons: InlineKeyboardButton): List<InlineKeyboardButton> {
        return ArrayList(Arrays.asList(*buttons))
    }

    fun removeCustomKeyboard(): ReplyKeyboard {
        return ReplyKeyboardRemove()
    }
}
